sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -166,18 +167,27 @@ object InMemoryRelation {
       outputOrdering: Seq[SortOrder],
       statsOfPlanToCache: Statistics): InMemoryRelation = {
     val relation = InMemoryRelation(output, cacheBuilder, outputOrdering)
-    relation.statsOfPlanToCache = statsOfPlanToCache
+    relation.setStatsOfPlanToCache(statsOfPlanToCache)
     relation
   }
+
+  val STATS_OF_PLAN_TO_CACHE_TAG = new TreeNodeTag[Statistics]("stats_of_plan_to_cache")
 }
 
 case class InMemoryRelation(
     output: Seq[Attribute],
     @transient cacheBuilder: CachedRDDBuilder,
     override val outputOrdering: Seq[SortOrder])
   extends logical.LeafNode with MultiInstanceRelation {
+  import InMemoryRelation.STATS_OF_PLAN_TO_CACHE_TAG
+
+  def setStatsOfPlanToCache(statsOfPlanToCache: Statistics): Unit = {
+    setTagValue(STATS_OF_PLAN_TO_CACHE_TAG, statsOfPlanToCache)
+  }
 
-  @volatile var statsOfPlanToCache: Statistics = null
+  def getStatsOfPlanToCache(): Statistics = {
+    getTagValue(STATS_OF_PLAN_TO_CACHE_TAG).get
+  }
```

Comment:
Hmm, `statsOfPlanToCache` has `volatile` semantics. But making it as a `TreeNodeTag`, seems we don't preserve it?

Reply:
good point. tree node tag should be thread safe as well.